cdc_iterator.go

```diff
@@ -0,0 +1,243 @@
+// Copyright © 2024 Meroxa, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package mysql
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/conduitio-labs/conduit-connector-mysql/common"
+	"github.com/conduitio/conduit-commons/csync"
+	sdk "github.com/conduitio/conduit-connector-sdk"
+	"github.com/go-mysql-org/go-mysql/canal"
+	"github.com/go-mysql-org/go-mysql/mysql"
+	"github.com/go-mysql-org/go-mysql/schema"
+	mysqldriver "github.com/go-sql-driver/mysql"
+)
+
+type cdcIterator struct {
+	canal *canal.Canal
+	acks  *csync.WaitGroup
+	data  chan *canal.RowsEvent
+
+	canalRunErrC chan error
+	canalDoneC   chan struct{}
+
+	config cdcIteratorConfig
+}
+
+type cdcIteratorConfig struct {
+	tables      []string
+	mysqlConfig *mysqldriver.Config
+	position    sdk.Position
+	TableKeys   common.TableKeys
+}
+
+func newCdcIterator(ctx context.Context, config cdcIteratorConfig) (common.Iterator, error) {
+	c, err := common.NewCanal(config.mysqlConfig, config.tables)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create canal: %w", err)
+	}
+
+	sdk.Logger(ctx).Info().Msg("created canal")
+
+	iterator := &cdcIterator{
+		canal:        c,
+		acks:         &csync.WaitGroup{},
+		data:         make(chan *canal.RowsEvent),
+		canalRunErrC: make(chan error),
+		canalDoneC:   make(chan struct{}),
+		config:       config,
+	}
+
+	startPosition, err := iterator.getStartPosition(config)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get start position: %w", err)
+	}
+
+	go func() {
+		handler := &cdcEventHandler{
+			canalDoneC: iterator.canalDoneC,
+			data:       iterator.data,
+		}
+		iterator.canal.SetEventHandler(handler)
+
+		iterator.canalRunErrC <- iterator.canal.RunFrom(startPosition)
+	}()
+
+	return iterator, nil
+}
+
+func (c *cdcIterator) getStartPosition(config cdcIteratorConfig) (mysql.Position, error) {
+	if config.position != nil {
+		pos, err := common.ParseSDKPosition(config.position)
+		if err != nil {
+			return mysql.Position{}, fmt.Errorf("failed to parse position: %w", err)
+		}
+		if pos.Kind == common.PositionTypeSnapshot {
+			return mysql.Position{}, fmt.Errorf("invalid position type: %s", pos.Kind)
+		}
+
+		return pos.CdcPosition.Position, nil
```

Comment:
Btw, I noticed that `pos.CdcPosition.Position` is a `mysql.Position`. Honestly, I would rather create a type of our own, even if it's the same as `mysql.Position`, just to have control over how the position gets marshalled into JSON. Imagine that `mysql.Position` changes for some reason - it could break a pipeline if we update the connector and the connector can't correctly unmarshal the last position anymore.